src/esphomelib/time/ds1307.cpp

```diff
@@ -0,0 +1,105 @@
+#include "esphomelib/defines.h"
+
+#ifdef USE_DS1307_TIME
+
+#include "esphomelib/time/ds1307.h"
+#include "esphomelib/log.h"
+#ifdef ARDUINO_ARCH_ESP8266
+  #include "sys/time.h"
+#endif
+
+ESPHOMELIB_NAMESPACE_BEGIN
+
+namespace time {
+
+static const char *TAG = "time.ds1307";
+
+DS1307Time::DS1307Time(I2CComponent *parent, uint8_t address) : I2CDevice(parent, address) { }
+
+void DS1307Time::dump_config() {
+  ESP_LOGCONFIG(TAG, "DS1307 Time:");
+  LOG_I2C_DEVICE(this);
+  if (is_failed()) {
+    ESP_LOGE(TAG, "DS1307 failed!");
+  }
+}
+
+float DS1307Time::get_setup_priority() const {
+  return setup_priority::HARDWARE_LATE;
+}
+
+void DS1307Time::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up DS1307...");
+  if (!read_data()) {
+    mark_failed();
+  }
+}
+
+bool DS1307Time::read_data() {
+  if (!read_bytes(0, ds1307_.raw, sizeof(ds1307_.raw))) {
+    ESP_LOGE(TAG, "Can't read I2C data.");
+    return false;
+  }
+  if (ds1307_.reg.ch) {
+    ESP_LOGW(TAG, "Clock halted.");
+    return false;
+  }
+  struct tm t {0};
+  t.tm_year = ds1307_.reg.year   + 10 * ds1307_.reg.year_10  + 100;
+  t.tm_mon  = ds1307_.reg.month  + 10 * ds1307_.reg.month_10 - 1;
+  t.tm_mday = ds1307_.reg.day    + 10 * ds1307_.reg.day_10;
+  t.tm_hour = ds1307_.reg.hour   + 10 * ds1307_.reg.hour_10;
+  t.tm_min  = ds1307_.reg.minute + 10 * ds1307_.reg.minute_10;
+  t.tm_sec  = ds1307_.reg.second + 10 * ds1307_.reg.second_10;
+  struct timeval tv {::mktime(&t), 0};
+  timezone tz = {0, 0};
+  ::settimeofday(&tv, &tz);
+  time::EsphomelibTime time = now();
+  char buf[128];
+  time.strftime(buf, sizeof(buf), "%c");
+  ESP_LOGD(TAG, "Synchronized time: %s", buf);
+  ESP_LOGV(TAG, "Rate select: %0d", ds1307_.reg.rs);
+  ESP_LOGV(TAG, "Square-Wave Enable (SQWE): %0d", ds1307_.reg.sqwe);
+  ESP_LOGV(TAG, "Output Control (OUT): %0d", ds1307_.reg.out);
+  return true;
+}
+
+bool DS1307Time::write_data() {
+  if (!read_bytes(0, ds1307_.raw, sizeof(ds1307_.raw))) {
+    ESP_LOGE(TAG, "Can't read I2C data.");
+    return false;
+  }
+  time::EsphomelibTime time = now();
```

Comment:
`now()` returns the localized time. Shouldn't the UTC time be written to the clock?

At least `settimeofday` expects UTC time values AFAIK, so if you would call read/write repeatedly when in a non-UTC timezone, you would increment the hour each time.

Reply:
That needs to be checked indeed! Right now, I am in the Germany/Berlin TZ and the time appears not to be incrementing. But you are right; it should...

Reply:
This is odd. You should be right. I should write UTC to the H/W clock, but when I do that, the time is st back one hour every time I write and read back from the H/W clock. Will investigate further on this...